--- qmail-1.03/splogger.c	1998-06-15 03:53:16.000000000 -0700
+++ qmail-1.03-patched/splogger.c	2014-07-29 15:56:06.000000000 -0700
@@ -32,13 +32,13 @@
   if (bufpos) {
     buf[bufpos] = 0;
     if (flagcont)
-      syslog(priority,"%s+%s",stamp,buf); /* logger folds invisibly; GPACIC */
+      syslog(priority,"+%s",buf); /* logger folds invisibly; GPACIC */
     else {
       stamp_make();
       priority = LOG_INFO;
       if (str_start(buf,"warning:")) priority = LOG_WARNING;
       if (str_start(buf,"alert:")) priority = LOG_ALERT;
-      syslog(priority,"%s %s",stamp,buf);
+      syslog(priority,"%s",buf);
       flagcont = 1;
     }
   }
